object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/CalledMethodsPredicateEvaluator.java

```diff
@@ -1,31 +1,176 @@
+        theBlock.getStatements().get(0).asReturnStmt().getExpression().get();
+
+    return expressionToBooleanFormula(theExpression, booleanFormulaManager);
+  }
+
+  /**
+   * Convert from Javaparser expression to sosy_lab boolean formula (the SMT solver's format).
+   *
+   * @throws UnsupportedOperationException if an unexpected expression is encountered
+   */
+  private static BooleanFormula expressionToBooleanFormula(
+      com.github.javaparser.ast.expr.Expression theExpression,
+      BooleanFormulaManager booleanFormulaManager) {
+    if (theExpression.isNameExpr()) {
+      return booleanFormulaManager.makeVariable(theExpression.asNameExpr().getNameAsString());
+    } else if (theExpression.isBinaryExpr()) {
+      if (theExpression.asBinaryExpr().getOperator().equals(BinaryExpr.Operator.OR)) {
+        return booleanFormulaManager.or(
+            expressionToBooleanFormula(
+                theExpression.asBinaryExpr().getLeft(), booleanFormulaManager),
+            expressionToBooleanFormula(
+                theExpression.asBinaryExpr().getRight(), booleanFormulaManager));
+      } else if (theExpression.asBinaryExpr().getOperator().equals(BinaryExpr.Operator.AND)) {
+        return booleanFormulaManager.and(
+            expressionToBooleanFormula(
+                theExpression.asBinaryExpr().getLeft(), booleanFormulaManager),
+            expressionToBooleanFormula(
+                theExpression.asBinaryExpr().getRight(), booleanFormulaManager));
+      }
+    } else if (theExpression.isEnclosedExpr()) {
+      return expressionToBooleanFormula(
+          theExpression.asEnclosedExpr().getInner(), booleanFormulaManager);
+    } else if (theExpression.isUnaryExpr()) {
+      if (theExpression.asUnaryExpr().getOperator().equals(UnaryExpr.Operator.LOGICAL_COMPLEMENT)) {
+        return booleanFormulaManager.not(
+            expressionToBooleanFormula(
+                theExpression.asUnaryExpr().getExpression(), booleanFormulaManager));
+      }
+    }
+    throw new UnsupportedOperationException(
+        "encountered an unexpected type of expression in an "
+            + "@CalledMethodsPredicate expression: "
+            + theExpression
+            + " was of type "
+            + theExpression.getClass());
+  }
+
+  /**
+   * Evaluate the given expression if every String in {@code trueVariables} is replaced by "true" in
+   * the boolean formula.
+   *
+   * @param expression a boolean formula in Java format, as a String
+   * @param trueVariables the variables in the boolean expression to treat as "true"
+   * @return whether the expression evaluates to true in the context where only the variables in
+   *     trueVariables are true, and all other variables are false
+   */
+  public static boolean evaluate(String expression, Collection<String> trueVariables) {
+
+    for (String cmMethod : trueVariables) {
+      expression = expression.replaceAll("\\b" + cmMethod + "\\b", "true");
```

Comment:
Rather than these string replacements before parsing, would it be better to do the substitution after parsing?

Reply:
I don't understand the suggestion. The way that this method works is to convert all names in the input expression into either `true` or `false` literals (depending on whether they are in `trueVariables` or not), and then evaluate the resulting expression.

It's a somewhat moot point regardless, though, because I will need to rewrite this code anyway to avoid the dependency on `SpelExpressionParser` before we can move this code into the framework. I'll re-evaluate the question when I'm doing that work.